### PR TITLE
cl/persistence/base_encoding: avoid per-scalar heap allocs in Rabbits codec

### PR DESCRIPTION
**Pull request by AskAlexSharov**
## Problem

Heap profiling shows `encoding/binary.Write` accounting for **~10% of all allocations** (475M objects), and 100% of those come from `base_encoding.WriteRabbits`, called on the antiquary state-collection path (`beaconStatesCollector.collectActiveIndices`) that serializes validator index sets.

`binary.Write(w, order, uint64Value)` allocates a fresh `[]byte` for each scalar — the buffer escapes to the heap because it's handed to the `io.Writer` interface. `WriteRabbits` calls it ~2× per non-contiguous run, so serializing a large sparse index set produces hundreds of millions of tiny 8-byte allocations. `ReadRabbits` has the identical anti-pattern via `binary.Read`.

## Fix

Use a single reused 8-byte buffer with `binary.LittleEndian.PutUint64`/`Uint64` and `io.ReadFull` instead of letting `binary.Write`/`binary.Read` allocate per scalar.

## Results

`BenchmarkWriteRabbits` on a 100k-element sparse list (~200k scalar writes):

| | allocs/op | B/op | ns/op |
|---|---|---|---|
| before | 200,080 | 1,652,629 | 3,235,850 |
| after | 78 | 51,502 | 1,313,157 |

~2,565× fewer allocations (residual 78 are zstd internals), ~2.5× faster.

## Wire format: unchanged

The encoding is still a zstd frame of little-endian `uint64` scalars, byte for byte, so existing data reads back identically.

`TestRabbitWriteWireFormat` / `TestRabbitReadWireFormat` pin this against a golden payload (`[0,1,2,5]` → scalars `4,3,2,1`) hand-computed from the format itself rather than captured from either codec — so the assertion doesn't rely on the writer and reader having changed together. Both tests, plus the existing `TestRabbit` round trip, pass **unmodified against the pre-change codec on `main`**; that is what establishes byte-level compatibility.

## Behavior change: malformed input is now rejected

This is not a pure no-op refactor. `ReadRabbits` now validates what it decodes instead of silently returning a truncated or garbage list. It returns an error when:

- the header length overflows `int`
- the decoded element count doesn't match the header
- an active run exceeds the remaining declared elements — this bounds the allocation a corrupt `count` can force, rather than growing `out` far past the header and only noticing at the end
- `current + count` overflows the index space
- the stream ends while a trailing contiguous-run count is still expected (header-only payload). `WriteRabbits` always emits one last, so that shape is corruption
- extra run-length scalars remain after the run that completes the declared output — a trailing gap scalar, or an appended zero-length gap/active pair that doesn't change the decoded list. `WriteRabbits` stops right after the final contiguous run, so the reader now requires immediate EOF there

Each has a subtest in `TestRabbitReadMalformed`. Checked against `main`: every one of those six inputs is currently accepted **silently, with no error** — which is the strictness this PR adds. Truncated scalars mid-stream already errored before and still do (now surfacing `io.ErrUnexpectedEOF` from `io.ReadFull`); they're covered too.

Well-formed input produced by `WriteRabbits` is unaffected — `TestRabbit`, `TestRabbitEmpty` (the `length == 0` edge case) and the antiquary / `persistence/state` round-trip tests pass unchanged.

### CODE REVIEW

File: cl/persistence/base_encoding/rabbit.go [hunk 1]
```diff
@@ -68,35 +83,52 @@ func ReadRabbits(out []uint64, r io.Reader) ([]uint64, error) {
...
+	length, err := readNum()
+	if err != nil {
 		return nil, err
 	}
 
-	if cap(out) < int(length) {
-		out = make([]uint64, 0, length)
+	if length > math.MaxInt {
+		return nil, fmt.Errorf("rabbit: encoded length %d overflows int", length)
+	}
```
[the remark this answers is not in the document]
Comment (AskAlexSharov) [reply cl/persistence/base_encoding/rabbit.go]: Addressed in 35c24df — changed the guard to `length > uint64(math.MaxInt)`.

Note this was already correct before: `math.MaxInt` is an *untyped* constant, so `length > math.MaxInt` (length being `uint64`) compiles and behaves correctly on both 32- and 64-bit platforms — `go build` and `TestRabbit` pass on the prior form. The explicit `uint64(...)` cast just makes the intent unambiguous and mirrors the `int(length)` cast used right after the guard.

**Review by domiwei — COMMENTED** [review]
Production code looks correct, and the allocation improvement is verified.

Before merging, please add focused regression coverage for:

- truncated 8-byte scalars
- decoded length not matching the header
- active runs exceeding the declared remaining length
- `current + count` overflow
- an independently constructed legacy-compatible encoding

The current writer-to-reader round trip cannot prove byte-level compatibility because both sides changed together.

Please also update the PR description: malformed inputs now fail more strictly, so this is not entirely a "no behavior change" change.

File: cl/persistence/base_encoding/rabbit.go [hunk 2]
```diff
@@ -68,35 +83,52 @@ func ReadRabbits(out []uint64, r io.Reader) ([]uint64, error) {
...
-			current += count
-		} else {
-			current += count
 		}
+		current += count
 		active = !active
 	}
+	if uint64(len(out)) != length {
+		return nil, fmt.Errorf("rabbit: decoded %d elements, header declared %d", len(out), length)
+	}
```
[the remark this answers is not in the document]
Comment (AskAlexSharov) [reply cl/persistence/base_encoding/rabbit.go:144]: Good catch — addressed in b85e1603a8.

Verified the premise before acting on it: `WriteRabbits` is the only producer of this format in the tree (the sole callers are `beaconStatesCollector.collectActiveIndices` writing and `state_accessors`/`antiquary` reading), and it always emits the trailing contiguous-run count, so the scalar count after the header is always odd and `active` is always `false` at EOF. `ReadRabbits` now returns an error otherwise.

Both shapes you named are covered as subtests, and both confirmed red before the fix — returning **no error** on the old code, exactly as you described:

- `stream ends after the header` — the `length == 0` header-only payload
- `stream ends on a gap run` — trailing gap that doesn't affect output length

`TestRabbitEmpty` pins the legitimate empty encoding (`WriteRabbits(nil)` → `0, 0`) so the check can't over-tighten into rejecting it.

File: cl/persistence/base_encoding/rabbit.go [hunk 3]
```diff
@@ -68,35 +83,57 @@ func ReadRabbits(out []uint64, r io.Reader) ([]uint64, error) {
...
+	// WriteRabbits always emits a contiguous-run count last, so a stream that stops
+	// while one is still expected is truncated or has trailing junk.
+	if active {
+		return nil, errors.New("rabbit: stream ends without a trailing contiguous-run count")
+	}
 	return out, nil
```
[the remark this answers is not in the document]
Comment (AskAlexSharov) [reply cl/persistence/base_encoding/rabbit.go:150]: Good catch — addressed in 67357e307b. Since `len(out)` only grows on active runs and `WriteRabbits` stops right after the run that completes the output, `ReadRabbits` now requires immediate EOF at that point and returns `rabbit: trailing data after the final contiguous run` otherwise.

Your exact example is pinned as the new `trailing zero-length runs after the final contiguous run` subtest (valid stream + appended `0, 0` pair), confirmed red — accepted with no error — before the fix. Two existing subtests were adjusted for the earlier, stricter check: `stream ends on a gap run` now trips it (still an error, different message), and the index-overflow payload moved its overflowing gap to before the output completes so it still exercises the overflow path. The canonical empty encoding (`0, 0` from `WriteRabbits(nil)`) stays accepted — `TestRabbitEmpty` guards against over-tightening.

File: cl/persistence/base_encoding/rabbit_test.go [hunk 4]
```diff
@@ -18,11 +18,32 @@ package base_encoding
...
+	"github.com/klauspost/compress/zstd"
 	"github.com/stretchr/testify/require"
 )
 
+func BenchmarkWriteRabbits(b *testing.B) {
+	// sparse list with many non-contiguous runs -> many scalar writes
+	list := make([]uint64, 0, 100000)
+	for i := range uint64(100000) {
+		list = append(list, i*3)
+	}
```
[the remark this answers is not in the document]
Comment (AskAlexSharov) [reply cl/persistence/base_encoding/rabbit_test.go:35]: This compiles. Since Go 1.22 the range expression may be of any integer type, not just `int` (https://go.dev/ref/spec#For_range: "for an integer value n, the iteration values 0 through n-1"), and this repo requires Go 1.25+. Verified: `go test -bench BenchmarkWriteRabbits ./cl/persistence/base_encoding/` builds and runs this loop as-is.

The form is also required here, not merely allowed: main enabled the modernize `rangeint` analyzer in #22449, which flags the `for i := uint64(0); i < n; i++` counter form this loop previously used — commit 78c340e616 switched it to `range` precisely to keep `make lint` green. Reverting to a counter loop would fail lint. Keeping as-is.

**Review by FrogRocky — APPROVED** [review]

**Review by JkLondon — APPROVED** [review]
